### PR TITLE
Adding smart pointers and new features to ArrayBase and friends.

### LINKED ISSUE #51: ArrayBase pull request.
@breznak  I need your help.  I am trying to submit a pull request but having a bit of a problem.  Here is what I did:
1) Cloned the htm-community/nupic.cpp repository.  
2) Opened a new local branch "Array"
3) copied my source files for ArrayBase, Array, ArrayRef to the new branch.
4) Made a few changes so that it would work.  built and ran the unit tests.
5) committed the changes and pushed them to github.com/dkeeney/nupic.cpp Array branch
6) On htm-community/nupic.cpp repository, clicked the pull request button.  Tried to specify the pull request coming from github.com/dkeeney/nupic.cpp Array branch to htm-community/nupic.cpp Master branch.    But it will not let me select that combination.

Am I doing this all wrong?

### PR DESCRIPTION
**Pull request by dkeeney**
Ok, I think I have this right this time :)
I pulled from upstream repository (this one) to get the latest then created a new Array branch,  then modified the 7 files.

The net affect of this change is to put the buffer into a smart pointer.  If the buffer is allocated then the Array owns the buffer and it will be deleted when Array is deleted.  If the buffer is set, then someone else owns the buffer and it will not be deleted.  If the Array object is assigned then the buffer is passed without a copy.

Fixes #51 

The new features will be needed later.  They are things like forcing a copy of the buffer, making a copy with type conversion, an alternative ways to manipulate the buffer container.

### CODE REVIEW

**Comment by dkeeney** [comment]
Hmmm, The CI tests all failed in the Python unit tests.   I don't understand the cause.  It does not appear to be related to the Array class unless it has something to do with the swig not handling it correctly.  Not sure how to address this.

File: src/nupic/types/BasicType.hpp [hunk 1]
```diff
@@ -68,9 +70,19 @@ class BasicType {
   // of NTA_BasicType_Int32 is "Int32"
   static NTA_BasicType parse(const std::string &s);
 
+  // convert elements of one array to another array of the specified types.
```
Comment (breznak) [inline src/nupic/types/BasicType.hpp]: PS: I think there is an automated doc system, so next time please stick to the 
/**
 *
 */ for method doc comments. You can use goodies as @param @return @throws etc

File: src/test/unit/ntypes/ArrayTest.cpp [hunk 2]
```diff
@@ -257,18 +257,7 @@ TEST_F(ArrayTest, testBufferAllocation) {
...
-        a.allocateBuffer(10);
-      } catch (nupic::Exception) {
-        caughtException = true;
-      }
-
-      ASSERT_TRUE(caughtException)
-          << "Test case: " + testCase->first +
-                 " - allocating a buffer when one is already allocated should "
-                 "raise an exception";
+      // Note: reallocating a buffer is now allowed.  dek, 08/07/2017
```
Comment (breznak) [inline src/test/unit/ntypes/ArrayTest.cpp]: :+1:

File: src/test/unit/ntypes/ArrayTest.cpp [hunk 3]
```diff
@@ -299,18 +288,7 @@ TEST_F(ArrayTest, testBufferAssignment) {
...
-      a.setBuffer(buf2.get(), testCase->second.allocationSize);
-    } catch (nupic::Exception) {
-      caughtException = true;
-    }
-
-    ASSERT_TRUE(caughtException)
-        << "Test case: " + testCase->first +
-               " - setting a buffer when one is already set should raise an "
-               "exception";
+    // setting a buffer when one is already set is now allowed. dek 08/07/2018
```
Comment (breznak) [inline src/test/unit/ntypes/ArrayTest.cpp]: so what happens when one buff is already allocated and you set other?

**Comment by dkeeney** [comment]
>
> throw runtime err of negative element is being converted to unsigned? I
> understand we want to allow sign->unsign conversion.
>
Good idea. I thought about it at the time and then forgot to follow
through.  In most cases the data is either zero or non-zero. But still we
should check this.

PS: I think there is an automated doc system,
>
Sorry, got into bad habits under Windows which does not use that.

so what happens when one buff is already allocated and you set other?
>
The smart pointer will delete the old and create the new.  (will not leak)
Now, if someone gets the real buffer pointer using getBuffer() and then
reallocates, the original real pointer becomes invalid.  So we do need to
make sure that that when real pointers they are only used locally and don't
get passed around.  The Array object should be passed around instead.

**Comment by dkeeney** [comment]
Ok, I now have the range checks in the BaseType conversion routines.  Sorry it took so long, life is getting in the way :)
Having a bit of a problem keeping Visual Studio from changing the format of all lines. Not just the line endings but also the indent and layout.  I am going to try using beyond compare as a merge tool and see if that helps keep unchanged lines from being modified.

File: src/test/unit/types/BasicTypeTest.cpp [hunk 4]
```diff
@@ -1,108 +1,579 @@
...
+  //////////////////////// To Real64 //////////////////////////////////////////////////////////////
+
+  BasicType::convertArray(ca.dest, NTA_BasicType_Real64, ca.bufByte,
+                          NTA_BasicType_Byte, 8);
+  ASSERT_TRUE(ca.checkArray<Real64>(ca.dest)) << "Byte to Real64 conversion";
+  // Note: Bytes are char, not unsigned char.
+
+  BasicType::convertArray(ca.dest, NTA_BasicType_Real64, ca.bufInt16,
+                          NTA_BasicType_Int16, 8);
+  ASSERT_TRUE(ca.checkArray<Real64>(ca.dest)) << "Int16 to Real64 conversion";
```
Comment (breznak) [inline src/test/unit/types/BasicTypeTest.cpp]: Very nice to have these tests! 

I am missing one that tests the error behavior? ei converting <int>{-1,0,1,2} to UInt.

File: src/nupic/types/BasicType.cpp [hunk 5]
```diff
@@ -1,558 +1,567 @@
...
+
+  if (!isValid(t))
+    throw Exception(__FILE__, __LINE__,
+                    "BasicType::getName -- Basic type is not valid");
+
+  return names[t];
+}
+
+// gcc 4.2 requires (incorrectly) these to be defined inside a namespace
+namespace nupic {
```
Comment (breznak) [inline src/nupic/types/BasicType.cpp]: Not part of this PR, but since it popped up: would it make sence to remove this? as gcc 4.2 is now unsupported anyway. What would be a benefit of NOT having these in the namespace `nupic::` ?
